app/grandchallenge/reader_studies/views.py

```diff
@@ -1429,3 +1468,326 @@ def form_valid(self, form):
+                "instance": None,
+            }
+        )
+        if self.request.method in ("POST", "PUT"):
+            data = json.loads(self.request.body)
+            for key in data:
+                if key in ["order", "csrfmiddlewaretoken", "new_interfaces"]:
+                    continue
+                interface = ComponentInterface.objects.get(slug=key)
+                if interface.is_image_kind:
+                    data[key] = [data[key]]
+            kwargs.update(
+                {
+                    "data": data,
+                }
+            )
+        return kwargs
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        media = Media()
+        for widget in self.form_class._possible_widgets:
+            media = media + widget().media
+        context.update(
+            {
+                "reader_study": self.reader_study,
+                "form_media": media,
+            }
+        )
+        return context
+
+    def _process_new_interfaces(self):
+        data = json.loads(self.request.body)
+        new_interfaces = data["new_interfaces"]
+        validated_data = {}
+        errors = {}
+        for entry in new_interfaces:
+            interface = ComponentInterface.objects.get(pk=entry["interface"])
+            if interface.is_image_kind:
+                entry["value"] = [entry["value"]]
+            form = DisplaySetInterfacesCreateForm(
+                data=entry,
+                pk=None,
+                interface=interface.pk,
+                user=self.request.user,
+                reader_study=self.reader_study,
+                auto_id="%s",
+            )
+            if form.is_valid():
+                cleaned = form.cleaned_data
+                validated_data[cleaned["interface"].slug] = cleaned["value"]
+            else:
+                errors.update({entry["interface"]: form.errors["value"]})
+        if errors:
+            raise ValidationError(errors)
+        return validated_data
+
+    def create_display_set(self, data):
+        ds = DisplaySet.objects.create(reader_study=self.reader_study)
+        ds.order = data.pop("order")
+        ds.save()
+        for slug in data:
+            if not data[slug]:
+                # Field is not filled in the form
+                continue
+            interface = ComponentInterface.objects.get(slug=slug)
+            civ = ComponentInterfaceValue(interface=interface)
+            if interface.requires_file:
+                user_upload = data[slug]
+                transaction.on_commit(
+                    partial(
+                        add_file_to_display_set.apply_async,
+                        kwargs={
+                            "user_upload_pk": str(user_upload.pk),
+                            "interface_pk": str(interface.pk),
+                            "display_set_pk": str(ds.pk),
+                        },
+                    )
```

Comment:
Stick to one way of constructing these, you use a lambda above and a partial here. This construction is the easiest to debug as you get the function signature back when you inspect the callbacks:

```
on_commit(
  add_file_to_display_set.signature(
    kwargs={
      "user_upload_pk": str(user_upload.pk),
      "interface_pk": str(interface.pk),
      "display_set_pk": str(ds.pk),
    }
  ).apply_async
)
```

Reply:
Ah that should work as well. I changed this one to a partial because lambdas don't work well in for loops (which took me a little while to figure out).